### PR TITLE
feat(vue-app): support named views

### LINKED ISSUE #4100: Nuxt support for named views for extendRoutes config
### Version

[v2.1.0](https://github.com/nuxt.js/releases/tag/v2.1.0)

### Reproduction link

[https://codesandbox.io/s/jvjm62ll5v](https://codesandbox.io/s/jvjm62ll5v)

### Steps to reproduce

extend route in nuxt.config.js

```
router: {
    extendRoutes (routes, resolve) {
      let index = routes.findIndex(route => route.path == '/')
      const children = [{
        path: 'sidebar',
        components: {
          default: resolve(__dirname, 'components/Sidebar.vue'),
          two: resolve(__dirname, 'components/SidebarTwo.vue')
        }
      }]

      routes[index]['children'] = children 
      console.log(util.inspect(routes, false, null))
    } 
  },
```

add two components:
`/components/Sidebar.vue`
```
<template>
  <div>
    <h3>Sidebar Default</h3>
  </div>
</template>
<script>
```
`/components/SidebarTwo.vue`
```
<template>
  <div>
    <h3>Sidebar Two</h3>
  </div>
</template>
```

Add `<router-view />` and `<router-view name="two" />` to index.vue

```
<template>
  <section class="container">
    <div>
      <h1 class="title">
        Homepage
      </h1>
      <h2 class="subtitle">
        My top-notch Nuxt.js project
      </h2>
      <nuxt-link to="/">Home</nuxt-link>
      <nuxt-link to="/sidebar">Sidebar</nuxt-link>
      <router-view /> 
      <router-view name="two" /> 
    </div>
  </section>
</template>
```

### What is expected ?

When you go to `/sidebar` route it should render both the default `<router-view />` and the named view `<router-view name="two" />`

### What is actually happening?

Cannot read property 'options' of undefined
An error occurred while rendering the page. Check developer tools console for details.

`_nuxt/app.js`
line: 2159

```
if (to.matched.length < 2) {
    // scroll to the top of the page
    position = {
      x: 0,
      y: 0
    };
  } else if (to.matched.some(function (r) {
    return r.components.default.options.scrollToTop;
  })) {
    // if one of the children has scrollToTop option set to true
    position = {
      x: 0,
      y: 0
    };
  } // savedPosition is only available for popstate navigations (back button)
```

### Additional comments?

Can't register named views in extendRoutes in nuxt.config with `components: {}` property only can register children using `component` property

<div align="right"><sub><em>This bug report is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c7958">#c7958</a>)</em></sub></div>

### PR DESCRIPTION
**Pull request by karakum**
## Types of changes
- [x] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
This change have some side effect for extendRoutes config. If you want to use named views, route item must contain `chunkNames` for each named component(except `default`).
For example:
```js
  router: {
    extendRoutes(routes, resolve) {
      let index = routes[0].children.findIndex(route => route.name === 'index-child-id');
      let child = routes[0].children[index];
      child.components = {
        default: child.component,
        left: resolve(__dirname, 'components/childLeft.vue'),
      };
      child.chunkNames = {
        left: 'components/childLeft',
      };
      routes[0].children[index] = child;
    },
  },
```
If community can help resolve chunk name from component's path, it is would be great and there is no need to hardcode chunkNames manually.

Resolves: #4100

## Checklist:
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly. (PR: https://github.com/nuxt/docs/pull/1053)
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

**Comment by manniL** [comment]
Please fix the lint issues @karakum

File: packages/vue-app/template/router.js
Comment (galvez) [inline packages/vue-app/template/router.js]: @karakum I think `for (const k in route.components) {` work here just as well -- would you please change it? Makes for better readability. Similarly, avoid having such a long line like:

```js
      resMap += firstIndent + tab + 'default: ' + (splitChunks.pages ? _name : `() => ${_name}.default || ${_name}`)
```

That can be split across multiple lines:

```js
      if (splitChunks.pages) {
        resMap += firstIndent + tab + 'default: ' + _name
      } else {
        resMap += firstIndent + tab + 'default: ' + `() => ${_name}.default || ${_name}`
      }
```

Which, despite some repetition, helps keep code easier to understand and maintain.

File: packages/vue-app/template/router.js
Comment (galvez) [inline packages/vue-app/template/router.js]: @karakum Following the same line of thought:

```js
        const cmpnnt = { _name, component: route.components[k] }
        if (k === 'default') {
          components.push({
          	...cmpnnt,
          	name: route.name,
          	chunkName: route.chunkName
          	})
        } else {
          components.push({
          	...cmpnnt,
          	name: `${route.name}-${k}`,
          	chunkName: route.chunkNames[k]
          })
          if (splitChunks.pages) {
          	resMap += `${nextIndent}${tab}${k}: ${_name}`
          } else {
          	resMap += `${nextIndent}${tab}${k}: ` + `() => ${_name}.default || ${_name}`
          }
        }
```

Anyway, just thought I'd encourage structuring code for readability. 

But PR looks great as-is. Cheers!

**Review by galvez — CHANGES_REQUESTED** [review]

**Comment by karakum** [comment]
@galvez, checkout requested changes

File: packages/vue-app/template/router.js
Comment (manniL) [inline packages/vue-app/template/router.js]: Just a nitpick: The variable name could be more expressive here :see_no_evil:

**Review by galvez — DISMISSED** [review]

**Comment by manniL** [comment]
@karakum Great! 👍 

Could you some small tests (a test fixture which looks like the example you've provided + unit tests that the example is actually showing what is expected), so the feature won't be broken by accident in future builds?

**Comment by karakum** [comment]
@manniL done!

**Comment by karakum** [comment]
@manniL is there any problem?

**Comment by manniL** [comment]
@karakum Not from my side. Waiting for the other reviews ☺️

File: examples/named-views/pages/index.vue
Comment (Atinux) [inline examples/named-views/pages/index.vue]: Why not also supporting `<nuxt-child name="left"/>` here?

**Comment by Atinux** [comment]
Great work @karakum 

I added my review on the only thing left that bother me: using `<router-view name="left"/>` instead of `<nuxt-child name="left"/>` and `<nuxt name="left"/>` when used in layouts?

Of course we will need to update nuxt-child to support it but I think it's necessary in order to support this feature completely.

**Review by atinux — CHANGES_REQUESTED** [review]

File: examples/named-views/pages/index.vue
Comment (karakum) [reply examples/named-views/pages/index.vue]: You are right!

**Comment by Atinux** [comment]
Awesome 🙌 

Did you try with `<nuxt name="left">` too?
For this we need to update the parent component in the route to have two components as well.

**Comment by karakum** [comment]
@Atinux, no, I did not. I just not understand this use-case.

**Comment by karakum** [comment]
Now I realize. I'll try it..

**Comment by karakum** [comment]
@Atinux `<nuxt name="left"/>` implemented, with tests & example. Now it really awesome!

**Comment by galvez** [comment]
ping @Atinux @pi0 @clarkdo :)

**Review by atinux — APPROVED** [review]

**Comment by Atinux** [comment]
This is so great @karakum :)

The only thing left before merging is to add a PR to the docs, the two pages needed to be updated are:
- https://nuxtjs.org/api/components-nuxt
- https://nuxtjs.org/api/components-nuxt-child

**Comment by karakum** [comment]
@Atinux https://github.com/nuxt/docs/pull/1053

**Comment by Atinux** [comment]
Thank you very much for this PR @karakum :)

**Comment by hashinteractive** [comment]
I still don't think this is working as intended:

https://codesandbox.io/embed/jvjm62ll5v

Can anyone confirm?

I've extended the index route to include a "Sidebar" component and have `<nuxt-child name="sidebar" />` in the index.vue page but it's not rendering the component.

**Comment by karakum** [comment]
@hashinteractive, maybe this feature is not released yet :)
And your example has old version  `"nuxt": "^2.1.0"`

**Comment by manniL** [comment]
latest nuxt edge should have it

**Comment by aaronransley** [comment]
Just wanted to drop in and thank @karakum everyone on the core team for their work creating and reviewing this! It looks like a wonderfully flexible feature, and I'm excited to reach for it in future projects.

**Comment by manniL** [comment]
@karakum I can't. Soon™️ 🙈
Starting with 2.4, we will also have a timed release plan 😉 

PS: you can try it already on `nuxt-edge`

**Pull request merged**
